### DISCUSSION TITLE
v2.0 - fromRequest not working as intended (?)

### DISCUSSION THREAD

**Discussion by coreation** [spatie/laravel-data#173] [Q&A] [answered]
First off, absolutely love the package, as someone who builds data intensive web application, I was really wondering how other people got through their array/object mess if there's a ton of data objects floating around.

DTOs solve that, and this is a very big step up for anyone working with them, so many thanks!

Second, I couldn't wrap my head around the magical conversion of injecting a DTO class into a controller method. According to the documentation, the creation of the DTO should be made based on the properties in the request, or, if defined, based on the fromRequest method of the DTO. 

What I experience, even [with the "song title" example](https://spatie.be/docs/laravel-data/v2/as-a-data-transfer-object/request-to-data-object) , is that the fromRequest is called, but the code first tries to populate it via request properties anyway. 

This leads to the fromRequest missing its goal (is it a bug or is this intentional?) because any additional properties you want to put into the constructor of your DTO via the fromRequest will need be defined as nullable in the constructor, since it first tries to make the DTO based on the request properties.

So my question is, is this intended behaviour or is this a bug? If the latter, happy to help out!

**Comment by jskorlol** [comment]
I think I need this feature.

**Comment by devenjahnke** [comment]
I'm experiencing the same issue. I cannot get the fromRequest method to trigger when creating a DTO via injection into the controller.

**Reply by coreation** [reply]
That's odd, that's not what I experience. The `fromRequest` is triggered, but it tries to make the DTO object first anyway, based on the request properties and the properties defined in the constructor. Once that is done, it then tries to build one using the `fromRequest` method. That's the behaviour that I believe is not intended as you want to override that default behaviour.

**Comment by rubenvanassche** [comment, marked as answer]
Seems indeed like a bug but cannot replicate it. We have a test which tests this feature (https://github.com/spatie/laravel-data/blob/8a6a8afaa714bb2bc975cd8b328ad9ffa19f03a3/tests/RequestDataTest.php#L246-L272) and this works like expected:

- We provide properties through request
- a fromRequest method will map these properties and create the data object correctly

Could you provide me a test where this feature isn't working? Then we can figure out what's going wrong.

**Reply by coreation** [reply]
Hey @rubenvanassche sorry I was on holidays so I missed your reply. I could spend some time figuring out how you guys build up your tests, to do so however I'll have to poke your knowledge about the package first. To start off with a base line, this test does test what you listed in the two bullets. However, it does not test, I believe, that if you pass a Data class into your function, it will not first be constructed using the request properties and afterwards (re)constructed using the `fromRequest` method, which is the crux of the issue. The method `fromRequest` will be called, but it will first try to use the normal behaviour and that's not what your test is testing for I think?

I'm all up for contributing to open source, so my question would first be, wouldn't it make sense in this case to write a test that uses a class inheriting from the Data class instead of one that has been made using magic properties/magic methods? The testcase would then test if the constructor is not called before the fromRequest method is called. I think it's too little of a change to add this test via a new branch/PR so I hope the code(changes) below suffices.

```
    /** @test */
    public function failing_it_can_manually_override_how_the_data_object_will_be_constructed()
    {
        Route::post('/other-route', function (SimpleData $data) {
            return ['name' => $data->string];
        });

        $this->postJson('/other-route', [
            'not_a_string' => 'Rick',
        ])
            ->assertOk()
            ->assertJson(['name' => 'Rick']);
    }
```

```
    // Changes to the "SimpleData" class, adding a new method so the test above executed the "affected" fromRequest override logic
    public static function fromRequest(Request $request): self
    {
        return new self($request->input('not_a_string'));
    }
``` 

When you execute the unit test `failing_it_can_manually_override_how_the_data_object_will_be_constructed` it should say "string field is required" indicating that the constructor of the Data class is first called, before the `fromRequest` is called. (But that assumption might be the wrong). This makes it so that building your constructor with properties that are build using the `fromRequest` is not possible and is - as far as I understand the package - not intended behaviour.

I went looking for the cause, which was right before my holiday, I'll see what I can figure out today in the time given to me. I hope the above example provides enough information to give me some pointers to locate/fix the issue.

**Reply by rubenvanassche** [reply]
Hi @coreation,

I took a look at your code and this is not a bug. We actually do not call constructors of the data object (you can try this one out by putting a dump statement within the constructor it won't do anything). When using magic creation methods a shorter version of the pipeline is used (https://spatie.be/docs/laravel-data/v2/advanced-usage/pipeline at the bottom of the page). If you want to achieve this then you should add #[WithoutValidation] attributes to your properties since the validator always will try to validate these properties hence the error.

**Reply by coreation** [reply]
Hi @rubenvanassche Thanks for getting back in touch, I actually did a "info" in the dump and was frowning on the fact that it didn't do anything, so that explains things! So this means that for each property that I know is going to be passed to the constructor via a special `from` method, I should add the #[WithoutValidation] attribute? I'll give that a go, thanks for the feedback!

**Reply by rubenvanassche** [reply]
Yeah, we want to be sure that when we're injecting a data object it is validated. In this case the payload you're providing doesn't completely map onto the data object so we're throwing validation errors. Good luck!

**Reply by coreation** [reply]
Thanks ^^ The injection that you mean is then the injection in the controller function, correct? Which triggers the validation but fails because it can't find everything it needs from request object? Then afterwards the `fromRequest` method is called and all is well because I return the necessary data to the constructor in that override function?

**Reply by rubenvanassche** [reply]
That's completely correct!

**Reply by coreation** [reply]
Perfect, thanks so much for taking the time to reply and maintaining this package, it really packs a lot of value.

**Reply by coreation** [reply]
This answers my question @rubenvanassche , looking back at this though, doesn't it make sense to change this behaviour? I'd expect that if I make a DTO where I add my required properties in my constructor and overwrite the behaviour when it's being made in a request, that if I add inject it into a Controller, where "Request" is often injected, it uses the fromRequest method and skipping some validations on its own? Not putting work into your hands, I might want to pick this up as well, as I think that'd be more the behaviour someone would expect?

**Reply by rubenvanassche** [reply]
Hmm I don't think the `fromRequest` method should validate stuff that's something the package should do. As soon data enters the fromRequest method the only thing you probably should do is correctly map data. I'm not sure this is a change we want't to make. But maybe you can convince me with a good example 😄

**Reply by coreation** [reply]
Hi @rubenvanassche , sorry to get back at this so late. The main reason for this is that I added a "WithoutValidation" attribute to properties that needed mapping and that sort of solved my issue. 

My proposal was not to add validation to the fromRequest method, but rather point out that when a request is coming in, and you use the DTO as a typehint in a controller function, then you have to add a lot of "WithoutValidation" annotations if that requests doesn't follow the same naming strategy as your DTO properties.

I was coming up with some proposals, but none of them really solve the issue in an eloquent way, so I think I'll leave it at that and go with the "WithoutValidation" solution ;) Just a few pointers!

_"As soon data enters the fromRequest method the only thing you probably should do is correctly map data."_
That's correct! And that's what I do, the issue is that if you don't use WithoutValidation annotations on the properties that are mapped in the `fromRequest` function, then exceptions are thrown because it tries to build the DTO, but cannot find the properties in the request. And that's what I find odd, because I assumed it would know "I'm being constructed in a controller method, so let's use the `fromRequest` method" but that's not the case.

Anyhow, thanks for the elaborate rundown of things, glad I got to get some insights under the hood ;)
